File: stryker4jvm-mutator-scala/src/main/scala/stryker4jvm/model/IgnoredMutationReason.scala
package stryker4jvm.model

/** Reason why a mutator did not produce a mutant
  */
sealed trait IgnoredMutationReason {
  def explanation: String
}

/** A mutation was excluded through user configuration
  */
final case object MutationExcluded extends IgnoredMutationReason {
  def explanation: String = "Mutation was excluded by user configuration"
}

/** Weapon-regeX gave a failure when parsing a regular expression
  */
final case class RegexParseError(pattern: String, message: String) extends IgnoredMutationReason {
  def explanation: String =
    s"The Regex parser of weapon-regex couldn't parse this regex pattern: '$pattern'. Please report this issue at https://github.com/stryker-mutator/weapon-regex/issues. Inner error: $message"
}

File: stryker4jvm-core/src/main/java/stryker4jvm/core/model/IgnoredMutationReason.java
package stryker4jvm.core.model;

public interface IgnoredMutationReason {
    /**
     * @return The reason why the mutation is ignored.
     */
    String explanation();

    final class MutationExcluded implements IgnoredMutationReason {

        @Override
        public String explanation() {
            return "Mutation was excluded by user configuration";
        }
    }
}
